Refactor IELE semantics to separate out logic for initializing accounts/storage/code/blockhashes

dwightguth opened the pull request:
In order to prepare the way for the IELE semantics to be used as the VM for a Cardano testnet, we refactor the semantics so that we no longer access account state or blockhashes except via specific klabels which load the relevant account into the accounts cell. We also configure the semantics so that some rules can be different depending on whether we are executing the semantics as a standalone or in Mantis.

The following changes are included in this PR:
* add standalone and node tags to pandoc so that we can have different rules depending on whether we are running a Mantis VM server.
* refactor the activeAccounts map into a set and use strictness so that we don't need to separately maintain a boolean for whether an account is empty or not.
* refactor `<storage>` cell so that we use `#lookupStorage` to compute the value of a storage key not in the storage map.
* refactor `<code>` cell so that code for an account can be missing and be computed by `#lookupCode`.
* refactor `iele.balance`, and `iele.extcodesize` so that an account can be missing and computed by `#newAccount`.
* fix bug in gas model of `sstore` where refunds were awarded for growing storage instead of for shrinking it.
* make process .build/vm/iele-vm which executes JSON tests using ProtoBuf API
* update JSON tests to more thoroughly test outputs

yzhang90 commented on iele-node.md:
```diff
@@ -0,0 +1,108 @@
+    rule <k> #lookupStorage(ACCT, INDEX) => . ... </k>
+         <account>
+           <acctID>  ACCT                                                         </acctID>
+           <storage> STORAGE => STORAGE [ INDEX <- #getStorageData(ACCT, INDEX) ] </storage>
```
why updating the storage here? What is "#getStorageData" supposed to do?

dwightguth commented:
I have tested this by running make test and make iele-test-node, so the tests are passing both in standalone and in node format. This still does not include networking, but it is now ready for review.

@bmmoore @theo25 @yzhang90 if you could take a look that would be good.

I am going to get started tomorrow on integrating the networking code.

theo25 commented on iele.md:
````diff
@@ -406,6 +406,129 @@ Some instructions require an argument to be interpreted as an address (modulo 16
+-   `#lookupCode` loads the `<code>` cell with the code for a particular account.
+
+```{.k .uiuck .rvk .standalone .node}
+    syntax InternalOp ::= #lookupCode ( Int )
````
where are the rewrite rules for this in case of standalone?

dwightguth replied:
They are not necessary as lookupCode only applies if the code is .Contract, which is not a well formed contract and is therefore not actually ever the code of any account; we simply reserve it to refer to an account that the node has not yet lazily computed code for.

dwightguth commented:
@theo25 please re review

yzhang90 commented on ethereum.md:
````diff
@@ -17,14 +23,18 @@ module ETHEREUM-SIMULATION
 Some IELE commands take a specification of IELE state (eg. for an account or transaction).
 
-```{.k .uiuck .rvk}
+```{.k .uiuck .rvk .standalone }
````
There is no IELESimulation defined in .node version? but In .node version, the k cell in the configuration still accepts IELESimulation, right?

yzhang90 commented on ethereum.md:
````diff
@@ -86,7 +96,7 @@ To do so, we'll extend sort `JSON` with some IELE specific syntax, and provide a
 -   `finishTx` is a place-holder for performing necessary cleanup after a transaction.
 
-```{.k .uiuck .rvk}
+```{.k .uiuck .rvk .standalone .node}
````
If I understand it correctly, the entry point for the .node version should be `runVM`, why include startTx, loadTx, and etc in the .node version?

dwightguth replied:
I have disabled more of the code in node that is not needed.

yzhang90 commented on ethereum.md:
```diff
@@ -182,14 +196,14 @@ To do so, we'll extend sort `JSON` with some IELE specific syntax, and provide a
  // ----------------------------------
-    rule <k> #exception _ ~> #finishTx => #popCallStack ~> #popWorldState ~> #popSubstate ... </k>
-    rule <k> #revert _    ~> #finishTx => #popCallStack ~> #popWorldState ~> #popSubstate ~> #refund GAVAIL ... </k> <gas> GAVAIL </gas>       
+    rule <k> #exception STATUS ~> #finishTx => #popCallStack ~> #popWorldState ~> #popSubstate ~> #load %0 STATUS ... </k>
```
Is this change (loading STATUS to the register) not listed in the PR summary?

dwightguth replied:
updated

yzhang90 commented on iele-semantics-plugin/MANTIS.ml:
```diff
@@ -0,0 +1,35 @@
+
+module Cache = Caching.Make(World.InMemoryWorldState)
+
+let hook_getBalance c _ _ config _ = match c with
```
Could you kindly remind me what are the five arguments to the function hook_getBalance?

dwightguth replied:
arguments, klabel, return sort, configuration, fresh function. This is the signature of all hooks in the OCAML backend, but most of these are not used by these hooks.

yzhang90 commented on iele.md:
```diff
@@ -406,6 +406,129 @@ Some instructions require an argument to be interpreted as an address (modulo 16
+         <activeAccounts> ACCTS (.Set => SetItem(ACCT)) </activeAccounts>
+         <accounts>
+           ( .Bag
+          => <account>
```
If the ACCT need to be created, should the balance, code, nonce cell be initialized?

dwightguth commented on data.md:
````diff
@@ -5,7 +5,7 @@ IELE uses arbitrary-precision integers, and sometimes also bytes (8 bit words).
 Both are implemented using K's `Int`.
 
-```{.k .uiuck .rvk}
+```{.k .uiuck .rvk .standalone .node}
````
I don't really understand the lua code very well, if you're interested in this, can you make a patch?

dwightguth commented:
I believe I have addressed all the issues with the exception of the Lua one, which Brandon says he will make an issue for. Since we want the IOHK people to be able to run this, I am going to merge this, but if you wanted anything else to be included message me and I'll make a followup PR asap.
